<a href="https://colab.research.google.com/github/THESHAMIK/Real-Life-Projects/blob/main/Deep_NN/KHDS_AutoKeras_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Here we tried to re-visit AUTOKERAS & tried to optimize it using more epochs, max trails, loss, metrics, callbacks, objective, tuner etc as per below site:
https://autokeras.com/structured_data_regressor/

Done most of experiments with both X & y as SCALED.

https://machinelearningmastery.com/autokeras-for-classification-and-regression/ 



In [ ]:
# Import packages
import numpy as np
import pandas as pd
import random
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from keras.models import Sequential
from keras.layers import Dense, BatchNormalization, Dropout
#from keras.optimizers import SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from tensorflow.keras.optimizers import SGD, RMSprop, Adadelta, Adagrad, Adamax, Nadam, Ftrl
from tensorflow.keras.optimizers import Adam ##because cannot import name 'Adam' from 'keras.optimizers'
from keras.callbacks import EarlyStopping, ModelCheckpoint
#from keras.wrappers.scikit_learn import KerasClassifier
from keras.wrappers.scikit_learn import KerasRegressor
from math import floor
from sklearn.metrics import make_scorer, accuracy_score
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import r2_score
#from bayes_opt import BayesianOptimization
from sklearn.model_selection import StratifiedKFold
from keras.layers import LeakyReLU
LeakyReLU = LeakyReLU(alpha=0.1)
import tensorflow as tf

import warnings
warnings.filterwarnings('ignore')
pd.set_option("display.max_columns", None)

In [ ]:
df = pd.read_csv("khds_final_1.csv")
df['Product Sulphur'] = pd.to_numeric(df['Product Sulphur'],errors = 'coerce') ##==> This one works as pd.to_numeric works for convert a list, a series, an array, or a tuple to a numeric datatype
filtered_df = df[df['Product Sulphur'].notnull()]
df.dropna(axis=0)
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 711 entries, 0 to 710
Data columns (total 34 columns):
 #   Column                                     Non-Null Count  Dtype  
---  ------                                     --------------  -----  
 0   LABfeed Sulphur                            711 non-null    float64
 1   T1 : FRESH FEED FLOW TO KHDS KGPerHour     711 non-null    float64
 2   T1 : Reactor inlet temperature DEGC        711 non-null    float64
 3   T1 : Hydrogen make-up flow NM3perHour      711 non-null    float64
 4   T1 : Recycle hydrogen flow NM3perHour      711 non-null    float64
 5   T1 : REACTOR OUTLET Temp DEGC              711 non-null    float64
 6   T1 : REACTOR FEED INLET Pressure KGperCM2  711 non-null    float64
 7   T1 : Cold separator pressure KGperCM2      711 non-null    float64
 8   T1 : KERO FEED DENSITY KGperCM2            711 non-null    float64
 9   T2 : FRESH FEED FLOW TO KHDS KGPerHour     711 non-null    float64
 10  T2 : Reactor inlet tempera

In [ ]:
target_col = "Product Sulphur"
X = filtered_df.loc[:, filtered_df.columns != target_col]
y = filtered_df.loc[:, target_col]

In [ ]:
from sklearn import metrics
import pandas
from keras.models import Sequential
from keras.layers import Dense
from keras.wrappers.scikit_learn import KerasRegressor
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from keras.activations import linear
import tensorflow.keras as keras
from tensorflow.keras import regularizers
import matplotlib.pyplot as plt
from keras.models import Sequential
from tensorflow.keras.optimizers import Adam# define model
from keras.optimizer_v2.adam import Adam

In [ ]:
y = y.to_numpy()
y = y.reshape(-1,1)
#trainy = trainy.to_numpy()
#trainy = trainy.reshape(-1,1)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(y)
# transform training dataset
y_scaled = input_scaler.transform(y)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(X)
# transform training dataset
X_scaled = input_scaler.transform(X)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
y.shape

(711, 1)

In [ ]:
!pip install autokeras

     |████████████████████████████████| 162 kB 4.3 MB/s 
     |████████████████████████████████| 133 kB 59.4 MB/s 
     |████████████████████████████████| 462 kB 53.8 MB/s 


In [ ]:
import plotly.express as px
import autokeras as ak
from autokeras import StructuredDataRegressor
from sklearn.model_selection import train_test_split
from IPython.display import HTML

In [ ]:
earlystopper = EarlyStopping(monitor='loss', min_delta=0, patience=15, verbose=1, mode='auto') #patience reduced to 10 from 15
#estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0, shuffle=False)
#history_5_4 = lr_model_d.fit(trainX_scaled, trainy_scaled, batch_size=30, epochs = 500, validation_data=(testX_scaled, testy_scaled),shuffle = True, verbose = 0, callbacks = [earlystopper])
#history_5_4 = lr_model_d_1.fit(trainX, trainy_scaled, batch_size=30, epochs = 500, validation_data=(testX, testy_scaled),shuffle = True, verbose = 0, callbacks = [earlystopper])
#estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0, shuffle=False,callbacks=[EarlyStopping(monitor='val_loss', patience=20)])
#estimator = KerasRegressor(build_fn=baseline_model, epochs=100, batch_size=5, verbose=0, shuffle=False,callbacks=[earlystopper])

In [ ]:
# It tries 10 different models.
reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")
#reg = ak.StructuredDataRegressor(overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss", tuner='random')
# Feed the structured data regressor with training data.
reg.fit(X_scaled, y_scaled, epochs=200, callbacks=[earlystopper])

Trial 40 Complete [00h 00m 09s]
val_loss: 4.023104667663574

Best val_loss So Far: 2.275585651397705
Total elapsed time: 00h 19m 59s
INFO:tensorflow:Oracle triggered exit
Epoch 1/200
23/23 [==============================] - 1s 4ms/step - loss: 1.2016 - mean_squared_error: 1.2016
Epoch 2/200
23/23 [==============================] - 0s 3ms/step - loss: 1.6616 - mean_squared_error: 1.6616
Epoch 3/200
23/23 [==============================] - 0s 3ms/step - loss: 0.9121 - mean_squared_error: 0.9121
Epoch 4/200
23/23 [==============================] - 0s 4ms/step - loss: 0.8055 - mean_squared_error: 0.8055
Epoch 5/200
23/23 [==============================] - 0s 4ms/step - loss: 0.7894 - mean_squared_error: 0.7894
Epoch 6/200
23/23 [==============================] - 0s 4ms/step - loss: 0.7485 - mean_squared_error: 0.7485
Epoch 7/200
23/23 [==============================] - 0s 4ms/step - loss: 0.7397 - mean_squared_error: 0.7397
Epoch 8/200
23/23 [==============================] - 0s 3ms/step -

Trial 3 Complete [00h 00m 03s]
val_loss: 4.315784931182861

Best val_loss So Far: 3.6199324131011963
Total elapsed time: 00h 00m 33s
Epoch 100/100
23/23 [==============================] - 0s 3ms/step - loss: 0.2593 - mean_squared_error: 0.2593
reg = ak.StructuredDataRegressor(max_trials=3, overwrite=True) ==> These were params on first pass1.

Pass2 ::
Trial 3 Complete [00h 00m 21s]
val_loss: 2.3486874103546143

Best val_loss So Far: **2.3**486874103546143
Total elapsed time: 00h 00m 44s

reg = ak.StructuredDataRegressor(max_trials=3, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")

Pass 3: With callbacks & max_trails = 30 from 3::
Trial 30 Complete [00h 00m 29s]
val_loss: 2.5138509273529053

Best val_loss So Far: **2.303**161382675171
Total elapsed time: 00h 10m 55s

PASS 4: tuner='bayesian' added but max_trails = 15 from 25 (hyperband was worst)::
Trial 25 Complete [00h 00m 05s]
val_loss: 4.327514171600342

Best val_loss So Far: 2.7460367679595947

PASS 5: tuner ='random' max trail = 25
Trial 25 Complete [00h 00m 05s]
val_loss: 4.380385398864746

Best val_loss So Far: 2.49894642829895
Total elapsed time: 00h 07m 32s

SO far randim works best among all tuners = bayesian, hyperband, greedy.

pass6: same as pass5 except max trails removed. by default its 100 but why callback not owkring?
Trial 100 Complete [00h 00m 16s]
val_loss: 4.087159156799316

Best val_loss So Far: 2.495527505874634
Total elapsed time: 00h 26m 50s

pass 7: all same above pass6, except epoch =200 from 100 & max trail =40
Trial 40 Complete [00h 00m 10s]
val_loss: 4.36867094039917

Best val_loss So Far: 2.4152731895446777
Total elapsed time: 00h 14m 13s

pass 8: same as pass 7 witrh epoch =300 from 200
Trial 40 Complete [00h 00m 06s]
val_loss: 4.809711933135986

Best val_loss So Far: 3.2346599102020264
Total elapsed time: 00h 10m 55s

PASS 9:: Replicating pass 3 with max trails = 40, epochs =200 & no tuner used
Trial 40 Complete [00h 00m 09s]
val_loss: 4.023104667663574

Best val_loss So Far: **2.275**585651397705
Total elapsed time: 00h 19m 59s

reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")

pass 11: All same as pass 10 but split rati set to 0.1 than 0.2 above. Also earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto')--> monitor changed from 'loss' to mse

Trial 30 Complete [00h 00m 32s]
val_loss: 0.21225327253341675

Best val_loss So Far: **0.121**28239125013351



In [ ]:
# Predict with the best model.
predicted_y_scaled = reg.predict(X_scaled)

23/23 [==============================] - 0s 5ms/step


In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y_scaled,y_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y_scaled,y_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y_scaled,y_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y_scaled,y_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y_scaled,y_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.33888006070618804
Final score (MSE): 0.1148396955442297
Final score (MAE): 0.20871200024821696
Final score (R2): 0.8726204283480066
Final score (MAPE): 3.9401550287474296


Earlier even with both scaling result was (for all params being same):
Final score (MSE): 6.122893377762371
Final score (MAE): 0.8093436874593429
Final score (R2): -17.210139256045256
Final score (MAPE): 5.452198501882145

Now on same params result is ::
Before save score (RMSE): 0.5015663350450803
Final score (MSE): 0.25156878845055375
Final score (MAE): 0.2706559128012366
Final score (R2): **0.608**6809130830875
Final score (MAPE): 6.073575709195961

All params same yes, BUT only diff is using X(scaled),y(scaled) in totality without splitting. So results were good as model got more data for training itself but result good
Before save score (RMSE): 0.3977670902847305
Final score (MSE): 0.15821865811358096
Final score (MAE): 0.23472967040490217
Final score (R2): **0.81**51491010882141
Final score (MAPE): 3.8867089743956265

PASS3:  With callbacks & max_trails = 30 from 3 (but still with full X & no splitting)::

Before save score (RMSE): 0.3536620359655597
Final score (MSE): 0.12507683568330483
Final score (MAE): 0.23672821980790304
Final score (R2): **0.8761**266873177348
Final score (MAPE): 8.117779046966902

PASS4: Added tuner = 'bayesian' (hyper band worked very bad). Max trail = 25

Before save score (RMSE): 0.3860245383558159
Final score (MSE): 0.14901494421282077
Final score (MAE): 0.24365072890837666
Final score (R2): 0.7921041448849359
Final score (MAPE): 3.3612951371158633

pass 5: Added tuner = 'random' (hyper band worked very bad). Max trail = 25

Before save score (RMSE): 0.3779711353591584
Final score (MSE): 0.1428621791646912
Final score (MAE): 0.19481394413346476
Final score (R2): 0.8259287165772973
Final score (MAPE): 10.418282811744374

PASS 6:Same as pass 5 except max trails removed so its equal to 100
Before save score (RMSE): 0.34210636155473045
Final score (MSE): 0.11703676261621594
Final score (MAE): 0.18231880587676905
Final score (R2): 0.8588170049866046
Final score (MAPE): 35.44145571959353
No much different from max trails = 25

pass 7:all same as pass6 exept epoch=200 & max trail 40 from 100
Before save score (RMSE): 0.3545376196520452
Final score (MSE): 0.12569692374853828
Final score (MAE): 0.18252120878757355
Final score (R2): 0.8411879757138232
Final score (MAPE): 5.190196233708742

PASS 9:: Replicating pass 3 with max trails = 40, epochs =200 & no tuner used
Before save score (RMSE): 0.33888006070618804
Final score (MSE): 0.1148396955442297
Final score (MAE): 0.20871200024821696
Final score (R2): **0.872**6204283480066
Final score (MAPE): 3.9401550287474296

reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")

PASS 10: Look at huge bad scores when train/test split is used:: All other criteria is same as above pass 9 that is epoch=200, max trail=40, no tuner used.

Before save score (RMSE): 0.8593099630341421
Final score (MSE): 0.7384136125697386
Final score (MAE): 0.3347192538045948
Final score (R2): -1.1029342767960197
Final score (MAPE): 7.819765244431312

pass 11: All same as pass 10 but split rati set to 0.1 than 0.2 above. Also earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto')--> monitor changed from 'loss' to mse

Trial 30 Complete [00h 00m 32s]
val_loss: 0.21225327253341675

Best val_loss So Far: **0.121**28239125013351
Before save score (RMSE): 0.3482562101307629
Final score (MSE): 0.12128238789464207
Final score (MAE): 0.26273287179509514
Final score (R2): 0.6448275219970139
Final score (MAPE): 2.2944849171203074

pass 12:::same as pass 10 but monitor = mse than loss. scores same as pass 10
Trial 40 Complete [00h 00m 20s]
val_loss: 0.7478248476982117

Best val_loss So Far: 0.7034339308738708
Total elapsed time: 00h 11m 36s
Before save score (RMSE): 0.8387097153018569
Final score (MSE): 0.7034339865417218
Final score (MAE): 0.328256171613517
Final score (R2): -0.9619806941781535
Final score (MAPE): 4.187767267610768


In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_scaled, test_size=0.20, random_state=2021)

**Difference between validation split & validation data**
validation_split: The best model found would be fit on the entire dataset **including the validation data**.

validation_data:  The best model found would be fit on the training dataset **without the validation** data.

In [ ]:
earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto') #patience reduced to 10 from 15
reg = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")
reg.fit(X_train, y_train, validation_data= (X_test, y_test), epochs=200, callbacks=[earlystopper])

Trial 40 Complete [00h 00m 20s]
val_loss: 0.7478248476982117

Best val_loss So Far: 0.7034339308738708
Total elapsed time: 00h 11m 36s
INFO:tensorflow:Oracle triggered exit
INFO:tensorflow:Assets written to: ./structured_data_regressor/best_model/assets


In [ ]:
# Predict with the best model.
predicted_y = reg.predict(X_test)

5/5 [==============================] - 0s 3ms/step


In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y,y_test))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y,y_test)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y,y_test)
print("Final score (MAE): {}".format(score1))
#score2 = metrics.r2_score(predicted_y,y_test)
score2 = metrics.r2_score(y_test, predicted_y)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y,y_test)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.8387097153018569
Final score (MSE): 0.7034339865417218
Final score (MAE): 0.328256171613517
Final score (R2): 0.3096852143066814
Final score (MAPE): 4.187767267610768


Look at huge reduction of R2 score when train/test split is used though Scaled COMPARED to when no spilitting is used
Before save score (RMSE): 0.8593099630341421
Final score (MSE): 0.7384136125697386
Final score (MAE): 0.3347192538045948
Final score (R2): -1.1029342767960197
Final score (MAPE): 7.819765244431312



In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y_scaled, test_size=0.10, random_state=2021)

In [ ]:
earlystopper = EarlyStopping(monitor='mean_squared_error', min_delta=0, patience=15, verbose=1, mode='auto') #patience reduced to 10 from 15
reg_1 = ak.StructuredDataRegressor(max_trials=40, overwrite=True, loss="mean_squared_error",metrics=None, objective="val_loss")
reg_1.fit(X_train, y_train, validation_data= (X_test, y_test), epochs=200, callbacks=[earlystopper])

Trial 30 Complete [00h 00m 32s]
val_loss: 0.21225327253341675

Best val_loss So Far: 0.12128239125013351
Total elapsed time: 00h 09m 58s
INFO:tensorflow:Oracle triggered exit
INFO:tensorflow:Assets written to: ./structured_data_regressor/best_model/assets


In [ ]:
# Predict with the best model.
predicted_y = reg.predict(X_test)

3/3 [==============================] - 0s 4ms/step


In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y,y_test))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y,y_test)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y,y_test)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y,y_test)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y,y_test)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 0.3482562101307629
Final score (MSE): 0.12128238789464207
Final score (MAE): 0.26273287179509514
Final score (R2): 0.6448275219970139
Final score (MAPE): 2.2944849171203074


In [ ]:
auto_keras_model = reg_1.export_model()
print(type(auto_keras_model)) 

<class 'keras.engine.functional.Functional'>


In [ ]:
try:
    auto_keras_model.save("model_autokeras", save_format="tf")
except Exception:
    auto_keras_model.save("model_autokeras.h5")

INFO:tensorflow:Assets written to: model_autokeras/assets


In [ ]:
from tensorflow.keras.models import load_model

In [ ]:
loaded_model = load_model("model_autokeras", custom_objects=ak.CUSTOM_OBJECTS)

In [ ]:
predicted_y = loaded_model.predict(X_unseen_scaled)

In [ ]:
from sklearn import metrics
score = np.sqrt(metrics.mean_squared_error(predicted_y,y_unseen_scaled))
print(f"Before save score (RMSE): {score}")
# Measure MSE error.  
score = metrics.mean_squared_error(predicted_y,y_unseen_scaled)
print("Final score (MSE): {}".format(score))
score1 = metrics.mean_absolute_error(predicted_y,y_unseen_scaled)
print("Final score (MAE): {}".format(score1))
score2 = metrics.r2_score(predicted_y,y_unseen_scaled)
print("Final score (R2): {}".format(score2))
score3 = metrics.mean_absolute_percentage_error(predicted_y,y_unseen_scaled)
print("Final score (MAPE): {}".format(score3))
#score4 = metrics.mean_squared_log_error(predicted_y_scaled,testy_scaled)
#print("Final score (MSLE): {}".format(score4))
#score5 = metrics.log_loss(predicted_y_scaled,testy_scaled)

Before save score (RMSE): 4.4050792542843915
Final score (MSE): 19.404723236526728
Final score (MAE): 2.527393555993899
Final score (R2): -0.4088693349926624
Final score (MAPE): 3.684403254055156


In [ ]:
t = pd.Dataframe(y_unseen_scaled)

AttributeError: ignored

In [ ]:
t

array([[10.05802259],
       [ 7.10842282],
       [ 9.28773383],
       [ 0.39447106],
       [-0.107953  ],
       [-0.17953341],
       [-0.6941952 ],
       [-0.79783275],
       [-0.76415169],
       [-0.35382616],
       [-0.62293168],
       [-0.42696695],
       [-0.40182127],
       [-2.23884102],
       [-0.28552692],
       [ 0.96667557],
       [-0.89194199],
       [-0.68320467],
       [-0.39747204],
       [ 2.65164796],
       [12.06404189],
       [ 5.90377418],
       [ 9.47839982],
       [-0.43754583],
       [ 5.28918496],
       [-1.20570087],
       [ 3.07580998],
       [ 0.22099734],
       [-0.05016284],
       [-0.02151065],
       [-0.08047052],
       [ 0.84180959],
       [ 0.46319367],
       [ 2.92380434],
       [ 1.08643067],
       [ 2.92132005],
       [-0.20446928],
       [-0.10837807],
       [ 0.33069189],
       [-0.06390222],
       [-1.35648992],
       [-0.57539349],
       [-0.01365246],
       [-1.26946416],
       [-0.40773237],
       [-0

After saving & loading model with split = 0.1 & r2=0.64 - pass 10
Before save score (RMSE): 100036.98826900923
Final score (MSE): 10007399021.933891
Final score (MAE): 99759.53966346153
Final score (R2): -179.5118581603864
Final score (MAPE): 0.9999725206664197
But this was x unseen & y both UNSCALED. Lets scale them:

pass2:
Before save score (RMSE): 4.4050792542843915
Final score (MSE): 19.404723236526728
Final score (MAE): 2.527393555993899
Final score (R2): -0.4088693349926624
Final score (MAPE): 3.684403254055156
NOT GOOD RESULTS

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import numpy as np

lr_model = LinearRegression()
lr_model.fit(X_scaled, y_scaled)
#scores = cross_val_score(lr_model, X_scaled, y_scaled, scoring='neg_mean_squared_error', cv=5,)
#scores = cross_val_score(lr_model, X_scaled, y_scaled, scoring='neg_mean_squared_error', cv=5,)
scores = cross_val_score(lr_model, X_scaled, y_scaled, scoring='r2', cv=5)

In [ ]:
print("CV Scores for each fold :: ", scores, 'Mean of CV scores ::', np.mean(scores))

CV Scores for each fold ::  [-33.10834916  -2.17383626  -0.49123921  -0.45799046  -0.2006836 ] Mean of CV scores :: -7.28641973841322


In [ ]:
from sklearn.metrics import r2_score
pred = lr_model.predict(X_test)
score2 = metrics.r2_score(pred,y_test)
print("Final score (R2): {}".format(score2))

Final score (R2): -2.744527647545975


In [ ]:
lr_model.fit(X_train, y_train)
pred_1 = lr_model.predict(X_test)
score2 = metrics.r2_score(pred_1,y_test)
print("Final score (R2): {}".format(score2))


Final score (R2): -1.92019810115033


In [ ]:
from sklearn.model_selection import cross_validate
from sklearn.metrics import make_scorer


In [ ]:
#from sklearn import cross_validation, linear_model
from sklearn.model_selection import cross_val_score
#kfold = KFold(n_splits=10)
#results = cross_val_score(reg, X, y, cv=kfold, scoring=mean_squared_error)
#print("Baseline: %.2f (%.2f) MSE" % (results.mean(), results.std()))

#Single metric evaluation using cross_validate
#cv_results = cross_validate(reg, X, y, cv=3)
#sorted(cv_results.keys())
#cv_results['test_score']

#scores = cross_val_score(reg, X, y, scoring='neg_mean_squared_error', cv=5,)




In [ ]:
df_test = pd.read_csv("KHDS_TestData_12Apr.csv")
df_test['Product Sulphur'] = pd.to_numeric(df['Product Sulphur'],errors = 'coerce') ##==> This one works as pd.to_numeric works for convert a list, a series, an array, or a tuple to a numeric datatype
df_test['LABfeed Sulphur'] = pd.to_numeric(df['LABfeed Sulphur'],errors = 'coerce')
filtered_df_test = df_test[df_test['Product Sulphur'].notnull()]
filtered_df_test = df_test[df_test['LABfeed Sulphur'].notnull()]
df.dropna(axis=0)
filtered_df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 52 entries, 0 to 51
Data columns (total 34 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   LABfeed Sulphur               52 non-null     float64
 1   T1 : FRESH FEED TO 17VVHI101  52 non-null     float64
 2   T1 : REACTOR FEED             52 non-null     float64
 3   T1 : HYDROGEN MAKE UP         52 non-null     float64
 4   T1 : RECYCLE HYDROGEN         52 non-null     float64
 5   T1 : REACTOR OUTLET           52 non-null     float64
 6   T1 : REACTOR FEED.1           52 non-null     float64
 7   T1 : SEPARATOR   OVHD         52 non-null     float64
 8   T1 : KERO FEED DENSITY        52 non-null     float64
 9   T2 : FRESH FEED TO 17VVHI101  52 non-null     float64
 10  T2 : REACTOR FEED             52 non-null     float64
 11  T2 : HYDROGEN MAKE UP         52 non-null     float64
 12  T2 : RECYCLE HYDROGEN         52 non-null     float64
 13  T2 : RE

In [ ]:
target_col = "Product Sulphur"
X_unseen = filtered_df_test.loc[:, filtered_df_test.columns != target_col]
y_unseen = filtered_df_test.loc[:, target_col]

In [ ]:
X_unseen.shape

(52, 33)

In [ ]:
y_unseen.shape

(52,)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(X_unseen)
# transform training dataset
X_unseen_scaled = input_scaler.transform(X_unseen)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)

In [ ]:
y_unseen = y_unseen.to_numpy()
y_unseen = y_unseen.reshape(-1,1)
#trainy = trainy.to_numpy()
#trainy = trainy.reshape(-1,1)

In [ ]:
#from sklearn.preprocessing import StandardScaler
#import numpy as np
# created scaler for only X predictors
input_scaler = StandardScaler()
# fit scaler on training dataset
# fit scaler
input_scaler.fit(y_unseen)
# transform training dataset
y_unseen_scaled = input_scaler.transform(y_unseen)
# transform test dataset
#testX_scaled = input_scaler.transform(testX)